### DISCUSSION TITLE
Implementation of spectral boundary intergal and possible link to BOOST c++ library

### DISCUSSION THREAD

**Discussion by chunhuizhao478** [idaholab/moose#26820] [Q&A General]
Hi all, I would like to implement spectral boundary integral (see Appendix B in the reference article: https://onlinelibrary.wiley.com/doi/full/10.1002/nag.2865), which involves complex values, time integration, inverse laplace transform, etc, just as a startup question, does MOOSE currently have sufficient functionality to implement it? I know the BOOST c++ library has special functions that could be suitable here, but I wonder how should I include the library within MOOSE? Thanks in advance for any suggestions!

**Comment by lindsayad** [comment]
We do not currently support using complex numbers in our MOOSE systems. @cticenhour has looked into that but I always forget what the status of it is. libMesh certainly supports configuring with complex numbers but we just haven't propagated that support into MOOSE. If you were writing custom user objects or something then you could certainly do whatever you want with `<complex>`. 

The easiest way to bring in boost into MOOSE is to configure libmesh pointing at a boost installation with the configure argument `--with-boost[=ARG]` where you can supply an installation directory for boost if it's in a non-standard location

**Reply by cticenhour** [reply]
It has been so long now since I was able to work on that that I am going to have to think hard on it... As you implied, getting PETSc and libMesh built wasn't a problem, but I think propagating the fact that `Number` is how a complex quality through all of our typedefs and templated methods was just a huge lift. My memory tells me that I started that conversion process, and I think I got hung up on the Residual and Jacobian methods and how to properly populate the matrices. I'm still interested in doing the work, but I might need to apply for some funding to have time to do it. It'll become more important to have that capability in future fusion projects I think.

**Reply by cticenhour** [reply]
To echo something @lindsayad said that is more useful to @chunhuizhao478 for the moment - I used `<complex>` in many objects in the electromagnetics module to do complex calculations under the hood, and then had the object output the portion of the calculation I needed (imaginary or real) to the variable representing that portion of the calculation.

**Reply by chunhuizhao478** [reply]
Thanks @cticenhour and @lindsayad for your help! I wonder if @cticenhour could direct me some examples you have? Thanks again!

**Reply by cticenhour** [reply]
[VectorEMRobinBC](https://mooseframework.inl.gov/source/bcs/VectorEMRobinBC.html) ([source code link for current version](https://github.com/idaholab/moose/blob/7cc4aedd3b71997307aaaf7cbae1810fd32698eb/modules/electromagnetics/src/bcs/VectorEMRobinBC.C)) is one example object where we have to do this. More usage and object examples can be found through the https://mooseframework.inl.gov/modules/electromagnetics page.
